endpoint/retry.go

```diff
@@ -53,3 +57,70 @@ func NewExponentialBackoffPolicy(i, m uint, d time.Duration) RetryPolicy {
 		return time.Duration(p) * d, true
 	}
 }
+
+// NewIndefiniteRetryPolicy is the policy that regulates indefinite amount of
+// retry attempts. Similar to NewExponentialBackoffPolicy, this policy allows a
+// fixed number of immediate attempts and calculates the next delay
+// exponentially. However, contrary to NewExponentialBackoffPolicy, this policy
+// always returns true as the second return value denoting that the message
+// should be retried indefinitely.
+//
+// i is the number of immediate attempts. d is a multiplier for the backoff
+// duration. dmax is the maximum backoff duration. randomize is the
+// randomization factor used to randomize the returned duration within the
+// following range:
+//
+// [ d-(d*randomize), d+(d*randomize) ]
+//
+// If d exceeds dmax, dmax value is used to randomize duration.
+func NewIndefiniteRetryPolicy(i uint, d, dmax time.Duration, randomize float64) RetryPolicy {
+	// lastDelay is the last computed delay returned by this retry policy
+	var lastDelay time.Duration
+	return func(env InboundEnvelope, _ error) (time.Duration, bool) {
+		n := env.AttemptCount
+
+		// If the attempt count is unknown, always retry, but always use the
+		// maximum backoff period.
+		if n == 0 {
+			lastDelay = dmax
+			return randomizeDuration(dmax, randomize), true
+		}
+
+		// Retry immediately if we haven't yet exhausted the immediate attempt
+		// limit.
+		if n < i {
+			return 0, true
+		}
+
+		// lastDelay has already been set to dmax, randomize dmax and return
+		if lastDelay == dmax {
+			return randomizeDuration(dmax, randomize), true
+		}
+
+		// Calculate the next exponential value
+		p := math.Pow(
+			2,
+			float64(n-i), // number of non-immediate attempts
+		)
+
+		if delay := time.Duration(p) * d; delay < dmax {
+			lastDelay = delay
+			return randomizeDuration(delay, randomize), true
+		}
+		// If we ended up here, delay is already exceeding dmax value.
+		// In this case we set lastDelay to dmax to signify that from now on
+		// we will keep returning dmax.
+		lastDelay = dmax
+		return randomizeDuration(dmax, randomize), true
+	}
+}
+
+// randomizeDuration randomizes d based on the randomization factor. The
+// randomized duration is guranteed to be within the range of
+// [ d-(d*randomize), d+(d*randomize) ].
+func randomizeDuration(d time.Duration, randomize float64) time.Duration {
```

Comment:
IMO retry policy should not be responsible for applying jitter.  If we need to add this we can add it in the receiver, but it should not really be necessary here as there is no "race" from multiple consumers to handle the message when the retry period is reached.